src/Build/Instance/ProjectInstance.cs

```diff
@@ -1314,11 +1329,30 @@ ICollection<ProjectItemInstance> IItemProvider<ProjectItemInstance>.GetItems(str
         /// immutable if we are immutable.
         /// Only called during evaluation, so does not check for immutability.
         /// </summary>
-        ProjectPropertyInstance IEvaluatorData<ProjectPropertyInstance, ProjectItemInstance, ProjectMetadataInstance, ProjectItemDefinitionInstance>.SetProperty(string name, string evaluatedValueEscaped, bool isGlobalProperty, bool mayBeReserved)
+        ProjectPropertyInstance IEvaluatorData<ProjectPropertyInstance, ProjectItemInstance, ProjectMetadataInstance, ProjectItemDefinitionInstance>.SetProperty(string name, string evaluatedValueEscaped, bool isGlobalProperty, bool mayBeReserved, bool isEnvVar)
         {
             // Mutability not verified as this is being populated during evaluation
             ProjectPropertyInstance property = ProjectPropertyInstance.Create(name, evaluatedValueEscaped, mayBeReserved, _isImmutable);
             _properties.Set(property);
+
+            if (_trackPropertyReads)
+            {
+                if (isEnvVar)
+                {
+                    if (!_environmentVariables.Contains(name))
+                    {
+                        // If it's a new env var, track it.
+                        _environmentVariables.Add(name);
+                    }
+                }
+                else if (_environmentVariables.Contains(name))
+                {
+                    // If it's NOT an env var but the property has the same name,
+                    // then we're not dependent on the env var any more: remove it.
+                    _environmentVariables.Remove(name);
+                }
```

Comment:
I lack enough context here. But if this mean the given name was first tracked as an env var and then replaced with a property, could it happen that it was queried after it was added as an env var, but before it was replaced by a property?

If the assumption is that this never happens, maybe we should add some invariant assertion (maybe under debug if we think it will be expensive) that states that after a 
TrackPropertyRead, then we should never reach Remove(name)

Reply:
If setting a property overwrites a previous environment variable property we won't track reads of it ANY MORE. However, if reads happened prior to when it was overwritten, those reads ARE captured.

Reply:
Ah, got it. Makes sense!

---
In reply to: [293497304](https://github.com/microsoft/msbuild/pull/4437#discussion_r293497304) [](ancestors = 293497304)